### PR TITLE
new benchmark tool

### PR DESCRIPTION
**Pull request by themanforfree**

### CODE REVIEW

File: benchmark/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    env_logger::builder()
+        .filter_module("benchmark", log::LevelFilter::Info)
+        .init();
+
+    let args: Benchmark = Benchmark::parse();
+
+    let mut runner = CommandRunner::new(args);
+    let stats = runner.run().await?;
+
+    println!("{}", stats.summary());
```
Comment (rogercloud) [inline benchmark/src/main.rs:16]: Please use log as you've init it at the beginning of the function.

File: benchmark/src/runner.rs [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+            "  Average:      {:.4} secs\n",
+            self.avg.as_secs_f64()
+        ));
+        s.push_str(&format!("  Requests/sec: {:.2}\n", self.qps));
+        s
+    }
+
+    /// Calculate the histogram from the latencies.
+    #[allow(clippy::indexing_slicing)]
+    pub fn histogram(&self) -> String {
```
Comment (rogercloud) [inline benchmark/src/runner.rs:93]: Is there any histogram lib existed? If not, can we provide one?

File: benchmark/src/runner.rs [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+            if let Some(err) = result.error {
+                let entry = self.errors.entry(err).or_insert(0);
+                *entry = entry.overflow_add(1);
+                continue;
+            }
+            bar.inc(1);
+            stats.latencies.push(result.elapsed);
+        }
+        stats.total = start.elapsed();
+        stats.qps = stats.latencies.len() as f64 / stats.total.as_secs_f64();
```
Comment (rogercloud) [inline benchmark/src/runner.rs:280]: please use numeric_cast instead of as

File: scripts/benchmark.sh [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+OUTPUT_DIR="${WORKDIR}/out"
+REPEAT=1
+SERVERS=("172.20.0.2" "172.20.0.3" "172.20.0.4" "172.20.0.5")
+CLUSTER_PEERS=(
+    ""
+    "${SERVERS[2]}:2379 ${SERVERS[3]}:2379"
+    "${SERVERS[1]}:2379 ${SERVERS[3]}:2379"
+    "${SERVERS[1]}:2379 ${SERVERS[2]}:2379"
+)
+XLINE_TEST_CASE=(
```
Comment (rogercloud) [inline scripts/benchmark.sh:12]: Can we make the shell simpler as there're many duplications.

File: benchmark/src/runner.rs (hunk 3 above)
Comment (themanforfree) [reply benchmark/src/runner.rs:280]: numeric_cast cannot cast usize to f64

File: scripts/benchmark.sh (hunk 4 above)
Comment (themanforfree) [reply scripts/benchmark.sh:12]: Yes, but cannot quickly run the specified test_case

File: benchmark/src/main.rs (hunk 1 above)
Comment (themanforfree) [reply benchmark/src/main.rs:16]: We need to output these information to stdout, log is only used to observe the running status of the program

File: benchmark/src/runner.rs (hunk 2 above)
Comment (themanforfree) [reply benchmark/src/runner.rs:93]: I found a lib but can't output a graph like this

**Review by rogercloud — APPROVED** [review]

**Pull request merged**